Gems/SceneProcessing/Code/Source/Generation/Components/TangentGenerator/TangentGenerateComponent.cpp

```diff
@@ -191,95 +184,175 @@ namespace AZ::SceneGenerationComponents
-        if ((requiredSpaces.size() == 1 && requiredSpaces[0] == AZ::SceneAPI::DataTypes::TangentSpace::FromSourceScene) && fbxTangentData && fbxBitangentData)
-        {
-            return true;
-        }
+        AZ::SceneAPI::DataTypes::TangentSpace tangentSpace = GetTangentSpaceFromRule(scene, nodeIndex);
 
         // Find all blend shape data under the mesh. We need to generate the tangent and bitangent for blend shape as well.
         AZStd::vector<AZ::SceneData::GraphData::BlendShapeData*> blendShapes;
         FindBlendShapes(graph, nodeIndex, blendShapes);
 
-        // Generate all the tangent spaces we need.
-        // Do this for every UV set.
+        // Generate tangents/bitangents for all uv sets.
         bool allSuccess = true;
-        size_t uvSetIndex = 0;
-        while (uvData)
+        for (size_t uvSetIndex = 0; uvSetIndex < uvSetCount; ++uvSetIndex)
         {
-            for (AZ::SceneAPI::DataTypes::TangentSpace space : requiredSpaces)
+            AZ::SceneAPI::DataTypes::IMeshVertexUVData* uvData = FindUvData(graph, nodeIndex, uvSetIndex);
+            if (!uvData)
             {
-                switch (space)
+                AZ_TracePrintf(AZ::SceneAPI::Utilities::ErrorWindow, "Cannot generate tangents for uv set %zu as it cannot be retrieved.\n", uvSetIndex);
+                continue;
+            }
+
+            // Check if we had tangents inside the source scene file.
+            AZ::SceneAPI::DataTypes::IMeshVertexTangentData* tangentData = FindTangentData(graph, nodeIndex, uvSetIndex);
+            AZ::SceneAPI::DataTypes::IMeshVertexBitangentData* bitangentData = FindBitangentData(graph, nodeIndex, uvSetIndex);
+
+            // If all we need is import from the source scene, and we have tangent data from the source scene already, then skip generating.
+            if ((tangentSpace == AZ::SceneAPI::DataTypes::TangentSpace::FromSourceScene))
+            {
+                if (tangentData && bitangentData)
                 {
-                // If we want Fbx tangents, we don't need to do anything for that.
-                case AZ::SceneAPI::DataTypes::TangentSpace::FromSourceScene:
+                    AZ_TracePrintf(AZ::SceneAPI::Utilities::LogWindow, "Using source scene tangents and bitangents for uv set %zu for mesh '%s'.\n",
+                        uvSetIndex, scene.GetGraph().GetNodeName(nodeIndex).GetName());
+                    continue;
+                }
+                else
                 {
-                    allSuccess &= true;
+                    // In case there are no tangents/bitangents while the user selected to use the source ones, default to MikkT.
+                    AZ_TracePrintf(AZ::SceneAPI::Utilities::WarningWindow, "Cannot use source scene tangents as there are none in the asset for mesh '%s' for uv set %zu. Defaulting to generating tangents using MikkT.\n",
```

Comment:
I think this should be a warning. It's likely user has a misunderstanding when selecting this option, or this rule is applied but the source asset has changed.

Reply:
Within the SceneAPI AZ_Warning and AZ_Error have (had?) a different meaning. To get reporting to work correctly AZ_TracePrintf should always be used but with specific windows (the first parameter). So in this case `AZ_TracePrintf(AZ::SceneAPI::Utilities::WarningWindow` is equivalent to calling AZ_Warning.

Reply:
Changed to AZ_Warning/AZ_Error and tested the effects, they show up as warnings/errors in the AP the same way. I was following the pattern that we use across the scene api with the traceprintfs.